src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -66,71 +64,273 @@ public void WriteMessage(HubMessage message, Stream output)
+                    {
+                        throw new InvalidDataException($"Unexpected JSON Token Type '{reader.TokenType}'. Expected a JSON Object.");
+                    }
 
-                    switch (type)
+                    do
                     {
-                        case HubProtocolConstants.InvocationMessageType:
-                            return BindInvocationMessage(json, binder);
-                        case HubProtocolConstants.StreamInvocationMessageType:
-                            return BindStreamInvocationMessage(json, binder);
-                        case HubProtocolConstants.StreamItemMessageType:
-                            return BindStreamItemMessage(json, binder);
-                        case HubProtocolConstants.CompletionMessageType:
-                            return BindCompletionMessage(json, binder);
-                        case HubProtocolConstants.CancelInvocationMessageType:
-                            return BindCancelInvocationMessage(json);
-                        case HubProtocolConstants.PingMessageType:
-                            return PingMessage.Instance;
-                        default:
-                            throw new InvalidDataException($"Unknown message type: {type}");
+                        switch (reader.TokenType)
+                        {
+                            case JsonToken.PropertyName:
+                                string memberName = reader.Value.ToString();
+
+                                switch (memberName)
+                                {
+                                    case TypePropertyName:
+                                        var messageType = reader.ReadAsInt32();
+
+                                        if (messageType == null)
+                                        {
+                                            throw new InvalidDataException($"Missing required property '{TypePropertyName}'.");
+                                        }
+
+                                        type = messageType.Value;
+                                        break;
+                                    case InvocationIdPropertyName:
+                                        invocationId = reader.ReadAsString();
+                                        break;
+                                    case TargetPropertyName:
+                                        target = reader.ReadAsString();
+                                        break;
+                                    case ErrorPropertyName:
+                                        error = reader.ReadAsString();
+                                        break;
+                                    case ResultPropertyName:
+                                        if (!reader.Read())
+                                        {
+                                            throw new JsonReaderException("Unexpected end when reading JSON");
```

Comment:
Wrap this reader + check for content in a helper method called CheckRead and reuse